app/src/main/java/com/davidtomas/taskyapp/features/agenda/data/agenda/repository/AgendaRepository.kt

```diff
@@ -0,0 +1,82 @@
+package com.davidtomas.taskyapp.features.agenda.data.agenda.repository
+
+import com.davidtomas.taskyapp.core.domain._util.Result
+import com.davidtomas.taskyapp.core.domain._util.getOrNull
+import com.davidtomas.taskyapp.core.domain._util.map
+import com.davidtomas.taskyapp.core.domain.model.DataError
+import com.davidtomas.taskyapp.features.agenda.data.agenda.remote.api.AgendaService
+import com.davidtomas.taskyapp.features.agenda.data.agenda.remote.mapper.toEventModel
+import com.davidtomas.taskyapp.features.agenda.data.agenda.remote.mapper.toReminderModel
+import com.davidtomas.taskyapp.features.agenda.data.agenda.remote.mapper.toTaskModel
+import com.davidtomas.taskyapp.features.agenda.data.event.local.source.EventLocalSource
+import com.davidtomas.taskyapp.features.agenda.data.reminder.local.source.ReminderLocalSource
+import com.davidtomas.taskyapp.features.agenda.data.task.local.source.TaskLocalSource
+import com.davidtomas.taskyapp.features.agenda.domain.model.AgendaItem
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.async
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.launch
+
+class AgendaRepository(
+    private val agendaService: AgendaService,
+    private val eventLocalSource: EventLocalSource,
+    private val reminderLocalSource: ReminderLocalSource,
+    private val taskLocalSource: TaskLocalSource
+) {
+    suspend fun getAgenda(): Result<List<AgendaItem>, DataError.Network> {
+        return coroutineScope {
+            if (returnAgendaItemsIfLocallySaved() != null) {
+                return@coroutineScope Result.Success(returnAgendaItemsIfLocallySaved()!!)
+            } else {
+                Result.Success(getFromRemoteSaveLocallyAndReturnItems())
+            }
+        }
+    }
+
+    private suspend fun CoroutineScope.returnAgendaItemsIfLocallySaved(): List<AgendaItem>? {
+        val events = async { eventLocalSource.getEvents() }.await().getOrNull()?.first()
+        val tasks = async { taskLocalSource.getTasks() }.await().getOrNull()?.first()
+        val reminders = async { reminderLocalSource.getReminder() }.await().getOrNull()?.first()
+        val listAgendaItem = mutableListOf<AgendaItem>()
+        if (!events.isNullOrEmpty()) events.forEach { listAgendaItem.add(it) }
+        if (!tasks.isNullOrEmpty()) tasks.forEach { listAgendaItem.add(it) }
+        if (!reminders.isNullOrEmpty()) reminders.forEach { listAgendaItem.add(it) }
+
+        return if (!listAgendaItem.isNotEmpty()) listAgendaItem else null
+    }
```

Comment:
Ah I see, in that case, I'd put the `coroutineScope` inside this function since this is the one that needs the child coroutines and not the parent one